blobstore.wit.md

```diff
@@ -0,0 +1,186 @@
+  list-containers: func() -> result<stream<container-name>, Error>
+
+  // retrieves a container by name
+  get-container: func(name: container-name) -> result<container, Error>
+
+  // deletes a container and all objects within it
+  delete-container: func(name: container-name) -> result<_, Error>
+
+  // returns true if the container exists
+  container-exists: func(name: container-name) -> result<bool, Error>
+
+  // copies (duplicates) an object, to the same or a different container.
+  // returns an error if the target container does not exist.
+  // overwrites destination object if it already existed.
+  copy-object: func(src: object-id, dest: object-id) -> result<_, Error>
+
+  // moves or renames an object, to the same or a different container
+  // returns an error if the destination container does not exist.
+  // overwrites destination object if it already existed.
+  move-object: func(src:object-id, dest: object-id) -> result<_, Error>
+}
+
+// a Container is a collection of objects
+resource container {
+
+  // returns container name
+  name: func() -> result<string, Error>
+
+  // returns container metadata
+  info: func() -> result<container-metadata, Error>
+
+  // begins reading an object
+  read-object: func(name: object-name) -> result<read-stream, Error>
+
+  // creates or replaces an object.
+  write-object: func(name: object-name) -> result<write-stream, Error>
+
+  // retrieves an object or portion of an object, as a resource.
+  // Start and end offsets are inclusive.
+  // Once a data-blob resource has been created, the underlying bytes are held by the blobstore service for the lifetime
+  // of the data-blob resource, even if the object they came from is later deleted.
+  get-data: func(name: object-name, start: u64, end: u64) -> result<data-blob, Error>
+
+  // creates or replaces an object with the data blob.
+  write-data: func(name: object-name, data: data-blob) -> result<_, Error>
+
+  // returns list of objects in the container. Order is undefined.
+  list-objects: func(name: object-name) -> result<stream<object-name>, Error>
+
+  // deletes object.
+  // does not return error if object did not exist.
+  delete-object: func(name: object-name) -> result<_, Error>
+
+  // deletes multiple objects in the container
+  delete-objects: func(names: list<object-name>) -> result<_, Error>
+
+  // returns true if the object exists in this container
+  has-object: func(name: object-name) -> result<bool, Error>
+
+  // returns metadata for the object
+  object-info: func(name: object-name) -> result<object-metadata, Error>
+
+  // removes all objects within the container, leaving the container empty.
+  clear: func() -> result<_, Error>
+}
+
+// A write stream for saving an object to a blobstore.
+interface "wasi:blob/write-stream" {
+
+  // writes (appends) bytes to the object.
+  write: func(data: list<u8>) -> result<_, Error>
+
+  // closes the write stream
+  close: func() -> result<_,Error>
```

Comment:
having close/flush here feels like it might be too low level. Ideally when I think about a "high level" blob store API, I don't want to have to worry about manually closing or flushing. Flushing, especially, feels like a lower level detail as I definitely don't want to have to remember when I do or do not have to flush.

Reply:
I agree. Here's my original thinking, for documentation:

Calling "flush" is never required. An app can correctly function and never call flush. Just like when editing a file, you don't need to hit ctrl-S or `:w` while you are still editing, but we still do it often. An app might have a connection open for a long time and want to minimize the risk of data loss or corruption in the event of a crash - that's when they use flush. 

An app should not have to worry about network buffering. A preferred implementation would let the "client" library issue flush periodically, or after an idle delay, like Nagle's algorithm does for tcp, then the client would never need to call flush. However, wasm modules aren't multithreaded or async, so it isn't easy to create a background task that could call flush. 

Removing flush from the component api puts the responsibility for smart buffering onto the service implementation, where it should be.